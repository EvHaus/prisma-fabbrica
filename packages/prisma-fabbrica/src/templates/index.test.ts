import { getDMMF } from "@prisma/internals";
import { template, printNode } from "talt";

import { modelScalarOrEnumFields, getSourceFile, findPrsimaCreateInputTypeFromModelName } from ".";

describe(modelScalarOrEnumFields, () => {
  it("generates literal type field type pattern", async () => {
    const dmmf = await getDMMF({
      datamodel: `
        model TestModel {
          id Int @id
          boolField Boolean
          strField String
          floatField Float
          dateTimeField DateTime
        }
      `,
    });
    const inputType = findPrsimaCreateInputTypeFromModelName(dmmf, "TestModel");
    const expected = template.sourceFile`
      type TestModelScalarOrEnumFields = {
        id: number;
        boolField: boolean;
        strField: string;
        floatField: number;
        dateTimeField: Date;
      }
    `();
    expect(printNode(modelScalarOrEnumFields("TestModel", inputType))).toBe(printNode(expected).trim());
  });

  it("does not generate for nullable field", async () => {
    const dmmf = await getDMMF({
      datamodel: `
        model TestModel {
          id Int @id
          nullableField Int?
        }
      `,
    });
    const inputType = findPrsimaCreateInputTypeFromModelName(dmmf, "TestModel");
    const expected = template.sourceFile`
      type TestModelScalarOrEnumFields = {
        id: number;
      }
    `();
    expect(printNode(modelScalarOrEnumFields("TestModel", inputType))).toBe(printNode(expected).trim());
  });
});

describe(getSourceFile, () => {
  it("generates TypeScript AST", async () => {
    const dmmf = await getDMMF({
      datamodel: `
        model User {
          id Int @id
          name String
        }
      `,
    });
    const sourceFile = getSourceFile({ document: dmmf, importSpecifierToPrismaClient: "@prisma/client" });
    expect(printNode(sourceFile)).toMatchSnapshot();
  });
});
