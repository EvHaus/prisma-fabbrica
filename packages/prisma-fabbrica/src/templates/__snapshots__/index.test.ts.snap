// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSourceFile generates TypeScript AST 1`] = `
"import { Prisma } from "@prisma/client";
import type { PrismaClient } from "@prisma/client";
import { getClient } from "@quramy/prisma-fabbrica";
import scalarFieldValueGenerator from "@quramy/prisma-fabbrica/lib/scalar/gen";
import { Resolver, resolveValue } from "@quramy/prisma-fabbrica/lib/helpers";
type UserScalarOrEnumFields = {
    id: number;
    name: string;
};
type UserFactoryDefineInput = {
    id?: number;
    name?: string;
};
type UserFactoryDefineOptions = {
    defaultData: Resolver<UserFactoryDefineInput>;
};
function autoGenrateUserScalarsOrEnums(): UserScalarOrEnumFields {
    return {
        id: scalarFieldValueGenerator.Int({ modelName: "User", fieldName: "id", isId: true, isUnique: false }),
        name: scalarFieldValueGenerator.String({ modelName: "User", fieldName: "name", isId: false, isUnique: false })
    };
}
export function defineUserFactory({ defaultData: defaultDataResolver }: UserFactoryDefineOptions) {
    const buildCreateInput = async (inputData: Partial<Prisma.UserCreateInput> = {}) => {
        const requiredScalarData = autoGenrateUserScalarsOrEnums();
        const defaultData = await resolveValue(defaultDataResolver);
        const defaultAssociations = {};
        const data: Prisma.UserCreateInput = { ...requiredScalarData, ...defaultData, ...defaultAssociations, ...inputData };
        return data;
    };
    const create = async (inputData: Partial<Prisma.UserCreateInput> = {}) => {
        const data = await buildCreateInput(inputData);
        return await getClient<PrismaClient>().user.create({ data });
    };
    return {
        _factoryFor: "User" as const,
        buildCreateInput,
        create,
    };
}
"
`;
